### ISSUE TITLE
Misleading warning in case of duplicate symbols in function's parameters declaration

### ISSUE THREAD

**Issue by DeLaGuardo** [clj-kondo/clj-kondo#1732]
**version**

clj-kondo v2022.05.31

**platform**

native on Mac OS and JVM version

**problem**

When by mistake the function's parameter declaration form contains identical symbols in different positions clj-kondo reports about "Unused binding" referring to all but the last such symbols. Clojure does not complain about that and the function will be compiled correctly but potentially it might cause problems which kondo can highlight. However, the current message might bring some confusion.

```clojure
(defn foo [x x] ;; => warning: unused binding x
  (inc x))
```

The same problem happens with destructuring used together with `:as` directive.

Original discussion begins in [slack thread](https://clojurians.slack.com/archives/C053AK3F9/p1655839370235429)

**repro**

```shell
echo '(defn foo [x x] (+ x x))' | clj-kondo --lint -
```

**expected behaviour**

The message could be different stating that there are identical bindings in parameters which will shadow each other.

I'm interested in contributing a PR if this issue seems relevant.

**Comment by borkdude** [comment]
I think we should warn on shadowed bindings (unless the name is `_`).

**Comment by borkdude** [comment]
But we have to think a about the expectations when we would use the name `:shadowed-binding`. E.g is `(let [x 1 x 2] ...)` also a shadowed binding? This is a common idiom and we should not warn against that, probably. So maybe another name would be better.

**Comment by DeLaGuardo** [comment]
> But we have to think a about the expectations when we would use the name `:shadowed-binding`. E.g is `(let [x 1 x 2] ...)` also a shadowed binding? This is a common idiom and we should not warn against that, probably. So maybe another name would be better.

no, this problem is only about parameters in function declaration where this is ~99% a mistake.

> I think we should warn on shadowed bindings (unless the name is `_`).

Consider also such situation:

1. Given a function `(defn foo [x x] ...)`
2. Then the developer decides to refactor via automatic rename of the var to `_x`
3. Warning is gone but the problem is still there

**Comment by borkdude** [comment]
Yes, but the name `_` is very common for ignored bindings, I'm not talking about `_x`. E.g. `(fn [_ _] ...)`

It could be interesting to "grasp" your local .m2 folder to see how common that is:

https://github.com/borkdude/grasp

**Comment by DeLaGuardo** [comment]
> Yes, but the name `_` is very common for ignored bindings, I'm not talking about `_x`. E.g. `(fn [_ _] ...)`
> 
> It could be interesting to "grasp" your local .m2 folder to see how common that is:
> 
> https://github.com/borkdude/grasp

Agree, even without grasp I know for sure I have hundreds of occurrences of `_` in my code :)

**Comment by borkdude** [comment]
@DeLaGuardo But also with double `_` in one argvec? This is what I meant :)

**Comment by DeLaGuardo** [comment]
> @DeLaGuardo But also with double `_` in one argvec? This is what I meant :)

I have a few, mostly from component definitions for integrant:

```clojure
(defmethod ig/init-key ::non-configurable-component [_ _] ...)
```

**Comment by borkdude** [comment]
Yeah, so I think we should ignore those, but warn on the rest.

**Comment by borkdude** [comment]
Keep

**Comment by borkdude** [comment]
perhaps a new rule `:same-named-param` or so would be good. Improvements for a name welcome.
